osu.Game/Rulesets/Objects/Pooling/HitObjectEntryManager.cs

```diff
@@ -0,0 +1,115 @@
+        /// </remarks>
+        private readonly Dictionary<HitObjectLifetimeEntry, HitObject?> parentMap = new Dictionary<HitObjectLifetimeEntry, HitObject?>();
+
+        /// <summary>
+        /// Stores the list of entries managed by this <see cref="HitObjectEntryManager"/> for each hit object managed by this <see cref="HitObjectEntryManager"/>.
+        /// </summary>
+        private readonly Dictionary<HitObject, List<HitObjectLifetimeEntry>> childrenMap = new Dictionary<HitObject, List<HitObjectLifetimeEntry>>();
+
+        public void Add(HitObjectLifetimeEntry entry, HitObject? parent)
+        {
+            if (parentMap.ContainsKey(entry))
+                throw new InvalidOperationException($@"The {nameof(HitObjectLifetimeEntry)} is already added to this {nameof(HitObjectEntryManager)}.");
+
+            var hitObject = entry.HitObject;
+            parentMap[entry] = parent;
+            entryMap[hitObject] = entry;
+
+            if (parent != null && childrenMap.TryGetValue(parent, out var parentChildEntries))
+                parentChildEntries.Add(entry);
+
+            hitObject.DefaultsApplied += onDefaultsApplied;
+
+            childrenMap[hitObject] = new List<HitObjectLifetimeEntry>();
+
+            OnEntryAdded?.Invoke(entry, parent);
+        }
+
+        public void Remove(HitObjectLifetimeEntry entry)
+        {
+            if (!parentMap.ContainsKey(entry))
+                throw new InvalidOperationException($@"The {nameof(HitObjectLifetimeEntry)} is not contained in this {nameof(HitObjectLifetimeEntry)}.");
+
+            var hitObject = entry.HitObject;
+            parentMap.Remove(entry, out var parent);
+            entryMap.Remove(hitObject);
+
+            if (parent != null && childrenMap.TryGetValue(parent, out var parentChildEntries))
+                parentChildEntries.Remove(entry);
+
+            hitObject.DefaultsApplied -= onDefaultsApplied;
+
+            // Remove all entries of the nested hit objects
+            if (childrenMap.Remove(entry.HitObject, out var childEntries))
+            {
+                foreach (var childEntry in childEntries)
+                    Remove(childEntry);
+            }
+
+            OnEntryRemoved?.Invoke(entry, parent);
+        }
+
+        public bool TryGet(HitObject hitObject, [MaybeNullWhen(false)] out HitObjectLifetimeEntry entry)
+        {
+            return entryMap.TryGetValue(hitObject, out entry);
+        }
+
+        /// <summary>
+        /// As nested hit objects are recreated, remove entries of the old nested hit objects.
+        /// </summary>
+        private void onDefaultsApplied(HitObject hitObject)
+        {
+            if (!childrenMap.Remove(hitObject, out var childEntries))
+                return;
+
+            foreach (var entry in childEntries)
+                Remove(entry);
+
+            childEntries.Clear();
+            childrenMap[hitObject] = childEntries;
```

Comment:
How is this safe? `OnDefaultsApplied` looks to be called after the new nested hitobjects have been created, not before. So if it's clearing `childEntries`, isn't it going to be removing entries of the new nested hitobjects?

Reply:
Just so I understand correctly, this seems to only be safe because the order of operations is:
- Hitobject changes.
- `HitObject.ApplyDefaults()
- `HitObject.DefaultsApplied()`
- `HitObjectEntryManager.onDefaultsApplied()`  (clears list)
- Sometime later...
- `DrawableHitObject.Apply()`
- `Playfield.GetPooledDrawableRepresentation()` (for each nested object)
- `HitObjectEntryManager.Add()` (adds to list

If I'm reading things correctly? I still find this pretty dodge and hard to follow, but I don't have any suggestions.